flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/CategoricalHistogram.scala

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.math
+
+import java.util
+
+/** Implementation of a discrete valued online histogram
+  *
+  */
+class CategoricalHistogram(
+    values: Array[Double],
+    counts: Array[Int] = Array.ofDim[Int](0))
+  extends OnlineHistogram
+  with Serializable {
+
+  // sanity checks
+  require(values.length > 0, "Number of classes should be a positive integer")
+  require(counts.length == 0 || counts.length == values.length, "Counts should have the same " +
+    "number" + " " + "of " + "entries as values")
+  val data = new util.ArrayList[(Double, Int)](values.length)
+  require(checkSanity, "All values in counts should be non-negative")
+
+  /** Number of bins currently being used
+    *
+    * @return number of bins
+    */
+  override def bins: Int = {
+    data.size
+  }
+
+  /** Adds a new instance
+    *
+    * @param p value to be added
+    */
+  override def add(p: Double): Unit = {
+    // search for the index where the value is equal to1 p
+    val search = find(p)
+    require(search >= 0, p + "is not present in the histogram")
```

Comment:
Yes. The categorical histogram is in fact not even mentioned in the paper. I saw fit to inherit them both from the same trait since effectively they represent statistics about a particular field, discrete or otherwise.